api/server/router/network/network_routes.go

```diff
@@ -0,0 +1,248 @@
+package network
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/api/server/httputils"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/pkg/parsers/filters"
+	"github.com/docker/libnetwork"
+	"github.com/docker/libnetwork/netlabel"
+)
+
+const (
+	byID = iota
+	byName
+)
+
+func (n *networkRouter) getNetworksList(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	if err := httputils.ParseForm(r); err != nil {
+		return err
+	}
+
+	filter := r.Form.Get("filters")
+	netFilters, err := filters.FromParam(filter)
+	if err != nil {
+		return err
+	}
+
+	list := []*types.NetworkResource{}
+	if names, ok := netFilters["name"]; ok {
+		for _, name := range names {
+			if nw, errRsp := findNetwork(n.netController, name, byName); errRsp == nil {
+				list = append(list, buildNetworkResource(nw))
+			}
+		}
+	} else if ids, ok := netFilters["id"]; ok {
+		for _, id := range ids {
+			// Return all the prefix-matching networks
+			l := func(nw libnetwork.Network) bool {
+				if strings.HasPrefix(nw.ID(), id) {
+					list = append(list, buildNetworkResource(nw))
+				}
+				return false
+			}
+			n.netController.WalkNetworks(l)
```

Comment:
I don't understand why we are not using this same `netController.WalkNetworks` function with a different parameter when we are filtering on name. This makes the two branches very different, and `findNetwork` happen to also deal with ids, so it's all quite confusing.